File: apps/app-frontend/src/helpers/pinned-items.ts
/**
 * Thin invoke() wrappers for the `pinned-items` Tauri plugin -- the sidebar
 * pin list, covering both client instances and locally hosted servers.
 */
import { invoke } from '@tauri-apps/api/core'

export type PinnedItemKind = 'instance' | 'hosted_server'

export interface PinnedItem {
	kind: PinnedItemKind
	ref_id: string
	position: number
}

export async function list(): Promise<PinnedItem[]> {
	return await invoke('plugin:pinned-items|pinned_items_list')
}

export async function pin(kind: PinnedItemKind, refId: string): Promise<void> {
	return await invoke('plugin:pinned-items|pinned_items_pin', { kind, refId })
}

export async function unpin(kind: PinnedItemKind, refId: string): Promise<void> {
	return await invoke('plugin:pinned-items|pinned_items_unpin', { kind, refId })
}

export async function setOrder(
	ordered: { kind: PinnedItemKind; ref_id: string }[],
): Promise<void> {
	return await invoke('plugin:pinned-items|pinned_items_set_order', { ordered })
}

File: apps/app/src/api/pinned_items.rs
use crate::api::Result;
use theseus::prelude::*;

#[tauri::command]
pub async fn pinned_items_list() -> Result<Vec<PinnedItem>> {
    Ok(theseus::pinned_items::list().await?)
}

#[tauri::command]
pub async fn pinned_items_pin(
    kind: PinnedItemKind,
    ref_id: String,
) -> Result<()> {
    Ok(theseus::pinned_items::pin(kind, ref_id).await?)
}

#[tauri::command]
pub async fn pinned_items_unpin(
    kind: PinnedItemKind,
    ref_id: String,
) -> Result<()> {
    Ok(theseus::pinned_items::unpin(kind, ref_id).await?)
}

#[tauri::command]
pub async fn pinned_items_set_order(
    ordered: Vec<PinnedItemRef>,
) -> Result<()> {
    Ok(theseus::pinned_items::set_order(ordered).await?)
}

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("pinned-items")
        .invoke_handler(tauri::generate_handler![
            pinned_items_list,
            pinned_items_pin,
            pinned_items_unpin,
            pinned_items_set_order,
        ])
        .build()
}
